feature/dim_student_surveys

jordan-springer opened the pull request:
<!--

NOTE: Please follow the instructions below when submitting a pull request
      For dbt Cloud Pull Request:
      * If you are trying to test your branch, submit a pull request and dbt Cloud will kick-off a run to validate.
      * If you are trying to **draft** a pull request, please select "Draft" so as to avoid dbt Cloud kicking-off and unnecessary job.
-->

# Description

Design two models to represent student survey data:
1. dim_curriculum_surveys
* design model to show survey structure and qa's

2. dim_curriculum_survey_responses
* model to associate student responses to surveys data 

## Follow-up Work
1. Review dim_course_structure to understand how to include contained_levels data 
2. Increment responses model to only capture new responses data (via user_levels)

## Links

- Spec: [Student Surveys](https://docs.google.com/document/d/1ZESOPi08kFVsF_fN2OkArB3hxPa4eLRDW5wKSl4dACI/edit?tab=t.0#heading=h.yadu993zas1t)
- Jira ticket(s): [1075](https://codedotorg.atlassian.net/browse/DATAOPS-1075)

## Testing story

Using the following survey as one example for validation:
https://studio.code.org/s/csd3-2024/lessons/28/levels/12/page/1

```/* testing script:
select 
    survey_name, 
    course_name, 
    survey_type, 
    question_type, question_number, question_text,
    answer_number, answer_text
from dbt_jordan.dim_curriculum_surveys
where script_name = 'csd3-2024' 
    -- and version_year = '2024'
    and survey_type = 'end of unit'
group by 
    1,2,3,4,5,6,7,8
order by question_number, answer_number asc 
```

- [ ] Does your change include appropriate tests on key columns?
      eg.
      - `not_null`
      - `unique`
      - `dbt_utils.unique_combination_of_columns: , ["value","value","value"...]

**Note: when submitting a new model for review please make sure the following have been tested:**

1. The model compiles (`dbt build -m 'your_model'`)
         or: _has the dbt Cloud job succeeded?_
4. The model runs (`dbt run -m 'your_model'`)
5. The model produces accessible data in the DW (`select 1 from 'your_model'`)

## Privacy

- [ ] 1.	Does this change involve the collection, use, or sharing of new Personal Data?
- [ ] 2.    Do these data exist in the appropriate schema(s)? 
- [ ] 3.	Does this change involve a new or changed use or sharing of existing Personal Data?
- [ ] 4.    Consider: will this data be visible on Tableau? will this data be surfaced in a report exported from Trevor?
- [ ] 5.    If yes to any of the above, please list the models, columns, and justification below:
      i.
      ii.
      iii. 


## PR Checklist:
--> **Note: if these are not all checked, the PR will be sent back.**

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code adheres to [style guide](https://docs.getdbt.com/best-practices/how-we-style/0-how-we-style-our-dbt-projects)👀 and is **[DRY](https://docs.getdbt.com/terms/dry)**
- [ ] Code is well-commented (**please do not leave extraneous commentary in model code, if it is for the purpose of documentation, please relocate accordingly)
- [ ] Appropriate documentation has been provided (see `.yml.`, did `dbt docs generate` succeed?)
- [ ] New features are translatable or updates will not break up/downstream models
- [ ] Relevant documentation has been added or updated (i.e. `dbt docs` has been updated successfully on [Github Pages](code-dot-org.github.io/analytics/)
- [ ] Pull Request is labeled appropriately (eg. `chore/`, `feature/`, `fix/`)
- [ ] Follow-up work items (including potential tech debt) are tracked and linked (if applicable)



<!--
changelog:
auth.      descr.      date
js         init        2024-01-22              
js         v1.2        2024-02-06
-->

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,89 @@
+course_structure as (
+    select * 
+    from {{ ref('dim_course_structure') }}
+    where course_name in ('csa','csd','csp')
```
```suggestion
```
We now have way more courses than csp, csa and csd, so this should include all student-facing content

nataliazm99 reviewed (changes requested):
The available responses for each questsion are missing

jordan-springer commented:
@nataliazm99 I am having trouble pulling in any level_types other than 'multi'. after tinkering with the joins, I think the `where level_name like 'survey%' is excluding rows that would be considered 'freeresponse' for example. there is also considerably less data than in your model as we are using `contained_levels` to drive this model. lmk your thoughts: 
![image](https://github.com/user-attachments/assets/2da9d150-0ce9-4f4a-95ec-7fa9aaa66210)

nataliazm99 commented:
The levels in your example are not survey questions, there are different activities in the curriculum where students are asked to enter free text answers to questions.
Here is the activity in the first level in your list: [link to activity](https://studio.code.org/s/k5howaimakesdecisions/lessons/2/levels/26/sublevel/1?section_id=5183664)
![image](https://github.com/user-attachments/assets/62dd62c5-f13e-49d1-b927-beaa33270bac)

nataliazm99 commented:
If you are missing FreeResponse levels it might be related to [this issue](https://codedotorg.slack.com/archives/C045UAX4WKH/p1733359449992759?thread_ts=1715098215.080409&cid=C045UAX4WKH) being investigated by the Teacher Tools team.

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,104 @@
+        cl.level_id                     as contained_level_id,
+        cl.level_name                   as question_name,
+        cl.level_type                   as question_type,
+        col.contained_level_position    as question_position,
```
This seems to be bringing only the page, not the order of the question in the survey. 
For the CSD pre-survey ([link](https://studio.code.org/s/csd1-2024/lockable/1/levels/1/page/1?&viewAs=Participant)) 
![image](https://github.com/user-attachments/assets/689ca8ba-c98c-4876-8c77-c4924e04ad78)
These three questions are in the first page, they should have positions 1,2,3. The current code returns 0. 
![image](https://github.com/user-attachments/assets/065078c8-23b0-4f82-af98-a9cb881ac785)

jordan-springer replied:
i looked into these and added `question_position` to the model. when you add that, you can see that positionally those zeroes are accurate and are other first questions on other pages: 


![image](https://github.com/user-attachments/assets/f5f7ccf6-6a37-4381-992a-a401f7d304e5)

nataliazm99 replied:
They match what is the properties field, but would be better to calculate the actual question number. You can do it with a   rank, ordering by question_page, question_position.

jordan-springer replied:
yeah let's calculate our own number to be more user-friendly. we could also just add a +1 to the position to make a new derived field like `question_position+1`

nataliazm99 replied:
Adding +1 won't work because the `question_position` value as pulled is restarting in every page (see values with page 0 have question position values 0-2, those with page 1 have question position 0-6, they should be 1-10)

nataliazm99 replied:
question_number is restarting every page, it should be consecutive for all questions in the same survey. 

For example, in the CSD pre-survey ([link](https://studio.code.org/s/csd1-2024/lockable/1/levels/1/page/2?&viewAs=Participant)) "I like computer science" is question N. 4. Because it is the first question in the second page, the `question_number` value is 1, when it should be 4. Use this survey for your validation.

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,104 @@
+        
+        cl.num_response_options,
+        cola.answer_text        as response_option,
+        cola.answer_number      as response_position,
```
```suggestion
        cola.answer_number,    
```
The answer_number is not the response position, it is the numerical value of the answer. For example, in a 7-point scale of Agree/Disagree 7 is the highest level of agreement, and 1 the lowest. This is used to calculate averages, so labeling it as the position would confuse users.

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,104 @@
+        col.contained_level_text        as question_text,
+        
+        cl.num_response_options,
+        cola.answer_text        as response_option,
```
```suggestion
        cola.answer_text,       
```
This way it corresponds to `answer_number`

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,89 @@
+
+combined as (
+    select 
+        lower(gl.level_name)            as survey_name,
```
```suggestion
        cs.content_area,
        cs.course_name,
        cs.script_id,
        cs.script_name,
        cs.unit, 
        cs.topic_tags,
        cs.version_year,
        col.level_group_level_id as survey_level_id,
        lower(gl.level_name)            as survey_name,
```
Reordering for ease of use

nataliazm99 commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,89 @@
+        col.contained_level_position    as question_position,
+        cl.num_response_options,
+        
+        cs.level_script_id,
```
Is this field necessary? If so, move after group level info

nataliazm99 reviewed (changes requested):
Left some comments and suggestions.

jordan-springer commented on dbt/models/marts/students/dim_student_surveys.sql:
```diff
@@ -0,0 +1,105 @@
+course_structure as (
+    select * 
+    from {{ ref('dim_course_structure') }}
+),
```
should this be content_area <> 'other'? is there a reason for the fuzzy match with `like`?

nataliazm99 replied:
like only works as a fuzzy match when using special characters, I prefer to use it with varchar fields, but you can decide to implement it in whatever way meets the need,

nataliazm99 reviewed (changes requested):
Change the logic for `question_number` to a dense_rank, partition only by `level_group_level_id` so it creates a consecutive numbering across all questions in a survey and doesn't restart on each page.

nataliazm99 reviewed (commented):
Looks good, let's wait until the `contained_levels` table is updated to approve so we can be sure it is working for other level types

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_surveys.sql:
I believe this code below would produce the same results with less complexity by leveraging course_structure and restricting records early on. I tested it with csd-post-survey-2023:
```
  with course_structure as (
select dcs.* 
from {{ ref('dim_course_structure') }}
    where content_area <> 'other'
    and level_name like '%survey%'
   )
--   
, contained_levels as (
    select *,
        dense_rank() over(
            partition by 
                level_group_level_id
            order by 
                contained_level_page, 
                contained_level_position asc) as question_number
    from {{ ref('stg_dashboard__contained_levels') }}
),
--
contained_levels_answers as (
    select distinct 
        level_id, answer_number, answer_text
    from {{ ref('stg_dashboard__contained_level_answers') }}
)
--
, levels as (
    select 
        *,   
        case 
            when level_type = 'multi' then json_array_length(json_extract_path_text(properties, 'answers'))
            when level_type = 'freeresponse' 	then 1 
            when level_type = 'external'		then 0 
        end 	as num_response_options 
--
    from {{ ref('stg_dashboard__levels') }}
)
--
   select 
  		cs.level_id                     as group_level_id,
        cs.content_area,
        cs.course_name,
        cs.script_id,
        cs.script_name,
        cs.unit, 
        cs.topic_tags,
        cs.version_year,
                col.level_group_level_id        as survey_level_id,
        lower(cs.level_name)            as survey_name,
        cs.level_script_id,
--
        case 
            when survey_name like '%pre%' 
             and survey_name not like '%preview%'  then 'pre'
            when survey_name like '%post%'         then 'post'
            when survey_name like '%pulse%'        then 'pulse'
            when survey_name like '%end of unit%'  then 'end of unit'
            else null end              as survey_type,
--
        col.contained_level_id,
        cl.level_name                   as question_name,
        cl.level_type                   as question_type,
        col.question_number,
        col.contained_level_text        as question_text,
--
        cl.num_response_options,
        cola.answer_text,       
        cola.answer_number
        from course_structure cs
--        
        join contained_levels   as col 
    	on cs.level_id  = col.level_group_level_id
--    	
    	join levels             as cl -- contained levels 
    	on col.contained_level_id = cl.level_id
--    	
    	left join contained_levels_answers as cola 
    	on cl.level_id = cola.level_id  
```

nataliazm99 reviewed (commented):
Proposing an alternative for dim_curriculum surveys

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_survey_responses.sql:
This model takes really long to build. Would it improve if 1) the level_surces staging model is materialized (I notice the user_levels and users ones are materialized) 2) the reference tables are filtered upon each other subsequently?
```
with 
student_surveys as (
    select * 
    from {{ ref('dim_curriculum_surveys') }}
),


user_levels as (
    select * 
    from {{ ref('stg_dashboard__user_levels' )}}
    where level_id in (select contained_level_id from student_surveys)
),


level_sources as (
    select * 
    from {{ ref('stg_dashboard_pii__level_sources') }}
    where level_source_id in (select level_source_id from user_levels)
),
```

nataliazm99 reviewed (commented):
The model has not finished building, it is at 4h:15min now, so here are some suggestions to improve performance. I will continue testing when it finishes building

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_surveys.sql:
```diff
@@ -16,113 +23,61 @@ contained_levels_answers as (
-        gl.level_id                     as group_level_id,
+, final as (
+   select 
+  		cs.level_id                     as group_level_id,
```
Either `group_level_id` or `survey_level_id` could be removed because they are the same. Probably better to remove group_level_id which is not used in `dim_curriculum_survey_responses`

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_surveys.sql:
```diff
@@ -16,113 +23,61 @@ contained_levels_answers as (
+    	
+    	left join contained_levels_answers as cola 
+    	on cl.level_id = cola.level_id  
+)
```
```{{ dbt_utils.group_by(19) }} --  grouping instead of select distinct to deduplicate records with better performance ```
Suggesting this because when testing I see records are duplicated. group by 19 if removing one of the group_level_id fields

Test case:
```
select 
*
 from dbt_natalia.dim_curriculum_surveys dcs 
where dcs.survey_name ilike 'Student End of Unit Survey_csd3_2024'
order by 
unit , question_number , answer_number 
```

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_survey_responses.sql:
Records are getting duplicated by multiples in this model too. Would need to make the change to `answer_texts` and group by{} the `combined` CTE to avoid duplicates.

Test case:
```
select 
student_id
, content_area
, course_name
, unit
, survey_level_id
, contained_level_id
, question_number
, question_text
, answer_number
, answer_response
, count(*)  n_records
 from dbt_natalia.dim_curriculum_survey_responses dcss 
where dcss.survey_name ilike 'Student End of Unit Survey_csd3_2024'
and student_id = 83794403
group by 1,2,3,4,5,6,7,8,9,10
order by 1,2,3,4,5,6,7,8,9,10
;
```

nataliazm99 reviewed (changes requested):
There are duplicates in both models, added comments with suggestions to avoid them.

jordan-springer commented:
@nataliazm99 
1. updated `dim_course_structure` to remove the `group_level_type` field which was causing duplication downstream.
2. updated responses model to remove _all_ answers and bring in only those selected. this was leading to fan out by a product of the number of available responses

jordan-springer commented:
Additional testing against tableau: 
```
select script_name, 
    course_name,
    question_number,
    question_text,
    answer_number,
    -- answer_response,
    count(distinct student_id)
from dbt_cloud_pr_518294_238.dim_curriculum_survey_responses
where course_name = 'csa'
    and script_name = 'csa8-2022'
    and survey_type = 'post'
    and school_year = '2022-23'
group by 1,2,3,4,5
order by question_number asc, answer_number asc
```

![image](https://github.com/user-attachments/assets/bfb5211b-22ff-4063-ac57-e202ccb9a9f1)

nataliazm99 commented on dbt/models/marts/students/dim_curriculum_surveys.sql:
Some of the question texts are coming in with text from the markdown:
![image](https://github.com/user-attachments/assets/b767d166-8fed-4145-9487-b218d4badea6)

 ```
 select * 
 from dbt_natalia.dim_curriculum_surveys dcs 
 where question_text ilike lower('%How challenging are the lessons and topics in this course for you?%')
 ;
 ```

jordan-springer replied:
This seems to be coming in from the contained_levels data itself:
```
select * 
from dbt_jordan.stg_dashboard__contained_levels 
where contained_level_id = 45967
```

jordan-springer replied:
So this could be something to resolve with the team doing this migration

nataliazm99 reviewed (commented):
I tested and data from analysis matches the survey answers model. Test story below. Some other comments here.
```
with
analysis as (
select
us.course_name_true
, us.script_name
, us.parent_level_id
, us.parent_level_name
, position
, questions
, item_type
, answer_number
, answer_text
, count(distinct user_id) n_students 
from dashboard.analysis.user_surveys us
where us.course_name_true = 'csp'
and us.survey_type = 'Post'
and us.script_name = 'csp-post-survey-2023'
and us.user_response_school_year = '2023-24'
--and us.script_level_school_year = '2023-24'
and answer_text is not null
and item_type = 'FreeResponse'
--and user_id = 106192580
group by 1,2,3,4,5,6,7,8,9
--order by 1,2,3,4,5,6,7
)
--
--
, dev as ( 
select
sr.course_name
, sr.script_name
, sr.survey_level_id
, sr.survey_name
, sr.question_number
, sr.question_text
, sr.question_type
, sr.answer_number
, sr.answer_response
, count(distinct sr.student_id) n_students 
from dbt_natalia.dim_curriculum_survey_responses sr 
where sr.course_name = 'csp'
and sr.survey_type = 'post'
and sr.script_name = 'csp-post-survey-2023'
and sr.school_year = '2023-24'
and sr.answer_response is not null
and question_type = 'freeresponse'
--and student_id = 106192580
--and us.script_level_school_year = '2023-24'
--and sr.us_intl = 'us'
group by 1,2,3,4,5,6,7,8,9
--order by 1,2,3,4,5,6,7
)
--
select a.* , d.*
from analysis a
full outer join dev d
on a.course_name_true 	= d.course_name
and a.script_name 		= d.script_name  
and a.parent_level_id	= d.survey_level_id
--and a.position			= d.question_number
--and lower(a.questions)	= d.question_text
and coalesce(a.answer_number,0)		= coalesce(d.answer_number,0)
and lower(a.answer_text) = lower(d.answer_response)
where  
--a.n_students is null
d.n_students is null
--coalesce(a.n_students,0) != coalesce(d.n_students,0)
order by a.position, d.question_number
;
```

jordan-springer commented on dbt/models/staging/dashboard/stg_dashboard__contained_level_answers.sql:
```diff
@@ -3,7 +3,7 @@ contained_level_answers as (
         level_id,
         answer_number,
-        answer_text,
+        lower(answer_text) as answer_text,
```
good callout on the `lower()`, I wouldn't apply a distinct in the staging layer to maintain what is true of the raw data
